Implement element-addressable memory

plafer opened the pull request:
Closes #1064

Note that I changed the `batch` column to hold `4 * previous_defn_of_batch` (now effectively the first address in the batch). This is to avoid a needless division per row in the processor.

This PR does not address the concern raised https://github.com/0xPolygonMiden/miden-vm/issues/1064#issuecomment-2549712514; we should do that one in a subsequent PR, as it is a backend thing, and getting this PR out the door will allow `miden-base` and the compiler to start working fixing their code.

plafer commented:
@bobbinth I am done implementing the processor part of the change, which is ready for review.

Left to do after (high-level):
- fix bus
- update air constraints
- update the book
- a bit of cleanup (the `TODO(plafer)`'s)

bobbinth commented on processor/src/chiplets/aux_trace/mod.rs:
```diff
@@ -919,26 +902,59 @@ fn build_memory_chiplet_responses<E>(main_trace: &MainTrace, row: RowIndex, alph
-    let value1 = main_trace.chiplet_memory_value_1(row);
-    let value2 = main_trace.chiplet_memory_value_2(row);
-    let value3 = main_trace.chiplet_memory_value_3(row);
+        alphas[0]
+            + alphas[1].mul_base(op_label)
+            + alphas[2].mul_base(ctx)
+            + alphas[3].mul_base(address)
+            + alphas[4].mul_base(clk)
```
nit: we do this computation in a few places - maybe makes sense to create a small helper function for it.

bobbinth commented on processor/src/chiplets/aux_trace/mod.rs:
```diff
@@ -919,26 +902,59 @@ fn build_memory_chiplet_responses<E>(main_trace: &MainTrace, row: RowIndex, alph
+        let idx0 = main_trace.chiplet_memory_idx0(row);
+        let idx1 = main_trace.chiplet_memory_idx1(row);
+
+        let value = if idx1 == ZERO && idx0 == ZERO {
+            main_trace.chiplet_memory_value_0(row)
+        } else if idx1 == ZERO && idx0 == ONE {
+            main_trace.chiplet_memory_value_1(row)
+        } else if idx1 == ONE && idx0 == ZERO {
+            main_trace.chiplet_memory_value_2(row)
+        } else if idx1 == ONE && idx0 == ONE {
+            main_trace.chiplet_memory_value_3(row)
+        } else {
+            panic!("Invalid batch indices. idx0: {idx0}, idx1: {idx1}");
+        };
```
nit: writing this as a `match` statement may look a bit nicer.

bobbinth commented on air/src/trace/chiplets/memory.rs:
```diff
@@ -1,53 +1,62 @@
-pub const MEMORY_INIT_READ: Selectors = [ONE, ZERO];
+// All bus labels encode the chiplet selector (1, 1, 0), as well as the read/write and element/word
+// columns. The purpose of the label is to force the chiplet to assign the correct values to the
+// read/write and element/word columns. We also include the chiplet selector as a "namespace" for
```
nit: I would probably use "domain separator" instead of "namespace" here.

bobbinth commented on air/src/trace/chiplets/memory.rs:
```diff
@@ -1,53 +1,62 @@
-/// The number of elements accessible in one read or write memory access.
-pub const NUM_ELEMENTS: usize = 4;
-
+/// Column to hold the whether the operation is a read or write.
+pub const READ_WRITE_COL_IDX: usize = 0;
```
nit: should we name this `READ_OR_WRITE` to be consistent with `ELEMENT_OR_WORD`?

bobbinth commented on assembly/src/assembler/mod.rs:
```diff
@@ -574,12 +574,14 @@ impl Assembler {
-            // - to allocate procedure locals we need to increment fmp by the number of locals
-            // - to deallocate procedure locals we need to decrement it by the same amount
-            let num_locals = Felt::from(num_locals);
+            // - to allocate procedure locals we need to increment fmp by 4 times the number of
+            //   locals
+            // - to deallocate procedure locals we need to decrement it by the same amount We leave
+            // 4 elements between locals to properly support reading and writing words to locals.
+            let locals_frame = Felt::from(num_locals * WORD_SIZE as u16);
```
nit: there is a period missing on line 579 and indentation missing on line 580.

Also, we should probably mention it very clearly in the docs that when we specify number of locals for a procedure, we are actually specifying the number of words. An alternative could be to specify the number of elements, but require them to be in multiples of 4 (this would be an easy compile-time error to catch).

bobbinth replied:
One thing we should do is make sure that the first procedure local is always word-aligned. Otherwise, it will be difficult to procedure code to know if they can use locals procedures which expect word-aligned memory pointers.

plafer replied:
Done.

> One thing we should do is make sure that the first procedure local is always word-aligned.

It is, since `FMP_MIN` is word-aligned, and procedures are only allowed to increment the `fmp` pointer by a multiple of 4

bobbinth commented on docs/src/user_docs/stdlib/collections.md:
```diff
@@ -15,7 +15,12 @@ The following procedures are available to read data from and make updates to a M
 | add         | Adds a new leaf to the MMR.<br /><br />This will update the MMR peaks in the VM's memory and the advice provider with any merged nodes.<br /><br />Inputs: `[N, mmr_ptr, ...]`<br />Outputs: `[...]`<br /><br />Where `N` is the leaf added to the MMR whose memory locations starts at `mmr_ptr`. |
 | pack        | Computes a commitment to the given MMR and copies the MMR to the Advice Map using the commitment as a key.<br /><br />Inputs: `[mmr_ptr, ...]`<br />Outputs: `[HASH, ...]`<br /><br /> |
-| unpack      | Load the MMR peak data based on its hash.<br /><br />Inputs: `[HASH, mmr_ptr, ...]`<br />Outputs: `[...]`<br /><br />Where:<br />- `HASH`: is the MMR peak hash, the hash is expected to be padded to an even length and to have a minimum size of 16 elements.<br />- The advice map must contain a key with `HASH`, and its value is `num_leaves \|\| hash_data`, and hash_data is the data u ...
+| unpack      | Writes the MMR who's peaks hash to `HASH` to the memory location pointed to by `mmr_ptr`.<br /><br />Inputs: `[HASH, mmr_ptr, ...]`<br />Outputs: `[...]`<br /><br />Where:<br />- `HASH`: is the MMR peak hash, the hash is expected to be padded to an even length and to have a minimum size of 16 elements.<br />- The advice map must contain a key with `HASH`, and its value is `[num_le ...
```
`mmr_ptr` should be word-aligned, no?

bobbinth reviewed (commented):
Looks good! Thank you! I reviewed most of the non-test and non-MASM code - though, I only had a superficial look at the constraint evaluation code. Left some comments inline - mostly small/nits.

A couple of questions:
- How did these changes impact performance?
- Would the current design support reading multiple elements from the same word in the same cycle? (see [here](https://github.com/0xPolygonMiden/miden-vm/issues/1610#issuecomment-2564669887) for context).

Al-Kindi-0 commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
```diff
@@ -490,19 +490,19 @@ export.compute_s1_norm_sq
         mem_loadw
 
-        # load pi_{i+512}
+        # load pi_{i+2048}
```
the index of the coefficient of the polynomial should stay the same

Al-Kindi-0 commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
```diff
@@ -490,19 +490,19 @@ export.compute_s1_norm_sq
         dup.12
         mem_loadw
-        #=> [PI, PI_{i+512}, C, pi_ptr, ...]
+        #=> [PI, PI_{i+2048}, C, pi_ptr, ...]
```
same comment here

Al-Kindi-0 commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
```diff
@@ -467,9 +467,9 @@ end
-#! The i-th coefficient s1_i of s1 is equal to c_i - (pi_i - pi_{512 + i}) which is equal to
-#! c_i  + pi_{512 + i} - pi_i. Now, we know that the size of the pi_i coefficients is bounded by
-#! J := 512 * q^2 and this means that J + pi_{512 + i} - pi_i does not Q-underflow and since
+#! The i-th coefficient s1_i of s1 is equal to c_i - (pi_i - pi_{2048 + i}) which is equal to
```
the index of the coefficient of the polynomial should stay the same

Al-Kindi-0 commented on stdlib/asm/crypto/stark/random_coin.masm:
We should probably add a disclaimer that most, if not all, of the procedures in this file assume that the input pointers are word-aligned.

Al-Kindi-0 commented:
Looks great, thank you!
I did an initial round of reviews and left some comments inline. Went through all files in detail except for the constraints one as it seems to have some todos.

plafer commented on CHANGELOG.md:
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+#### Highlights
+- [BREAKING] Memory is now memory addressable (#1598)
```
That would make more sense, yes. Fixed

plafer commented on processor/src/chiplets/memory/segment.rs:
```diff
@@ -166,41 +322,62 @@ impl MemorySegmentTrace {
+pub enum MemoryAccessType {
+    Element { addr_idx_in_batch: u8 },
+    Word,
+}
```
I considered doing this, but then we'd need to duplicate logic in the code that uses these methods [here](https://github.com/0xPolygonMiden/miden-vm/blob/6f3c11bc3e5ebfa8440397121503d8c88a47e348/processor/src/chiplets/memory/mod.rs#L323-L345). This is due to the fact that each variant of the proposed enum would be responsible for populating 2 columns (`read_or_write`, and `element_or_word`), but to handle a "read" or a "write" is the same for both words or elements (and similarly for element/word being the same independently of read or write). In pseudo-code, the code would go from

```
match operation {
    read => <read logic>,
    write => <write logic>
}
match access_type {
    element => <element logic>,
    word => <word logic>
}
```

to

```
match operation {
    read_word => {
        <read logic>,
        <word logic>
    },
    read_element => {
        <read logic>,   // duplicate
        <element logic>
    },
    write_word => {
        <write logic>,
        <word logic>   // duplicate
    },
    write_element => {
        <write logic>, // duplicate
        <element logic> // duplicate
    }
}
```

bobbinth commented on air/src/trace/mod.rs:
```diff
@@ -19,7 +19,7 @@ pub const MIN_TRACE_LEN: usize = 64;
 // ------------------------------------------------------------------------------------------------
 
 //      system          decoder           stack      range checks       chiplets
-//    (8 columns)     (24 columns)    (19 columns)    (3 columns)     (17 columns)
+//    (8 columns)     (24 columns)    (19 columns)    (3 columns)     (18 columns)
 // ├───────────────┴───────────────┴───────────────┴───────────────┴─────────────────┤
```
I think this sums up to 72 columns. Is this also what gets printed out when we run benchmarks? (I think we may be off by 1 somewhere as before this change the number of columns that gets printed out was 70).

plafer replied:
71 is the true count, the culprit was the range checks column (is actually 2 columns, not 3).

Fixed the comment

bobbinth commented on core/src/debuginfo/source_file.rs:
```diff
@@ -515,7 +515,7 @@ impl SourceContent {
     ///
     /// Returns `None` if the given index is out of bounds
     pub fn line_start(&self, line_index: LineIndex) -> Option<ByteIndex> {
-        self.line_starts.get(line_index.to_usize()).copied().map(ByteIndex::from)
+        self.line_starts.get(line_index.to_usize()).copied()
     }
```
Is this no longer needed?

plafer replied:
Indeed, the new `clippy` flagged it

bobbinth commented on miden/src/helpers.rs:
nit: we usually name such modules `utils.rs`.

plafer replied:
done

bobbinth commented on processor/src/chiplets/aux_trace/mod.rs:
```diff
@@ -971,7 +956,7 @@ where
 /// of alphas of matching length. This can be used to build the value for a single word or for an
 /// entire [HasherState].
 fn build_value<E: FieldElement<BaseField = Felt>>(alphas: &[E], elements: &[Felt]) -> E {
-    assert_eq!(alphas.len(), elements.len());
+    debug_assert_eq!(alphas.len(), elements.len());
     let mut value = E::ZERO;
     for (&alpha, &element) in alphas.iter().zip(elements.iter()) {
         value += alpha.mul_base(element);
```
I would add `#[inline(always)]` here.

Also, I wonder if the compiler would be able to unroll the loops here. Since we always know the length of `elements` at compile time, we could change this to look like so:
```rust
fn build_value<E: FieldElement<BaseField = Felt>, const N: usize>(alphas: &[E], elements: &[Felt; N]) -> E {
```
And then use `N` in the loop condition. Not sure if this will work, but at least there is enough info for the compiler to try to do loop unrolling.

plafer replied:
Changed it to your suggestion, except with `elements: [Felt; N]` (with the use of `N` explicitly in the loop just in case). It should indeed be enough for the compiler to unroll.

bobbinth commented on processor/src/chiplets/mod.rs:
```diff
@@ -286,91 +285,14 @@ impl Chiplets {
-        self.memory.write(ctx, addr, self.clk, new_word)?;
-
-        Ok(old_word)
+    /// Returns a reference to the Memory chiplet.
+    pub fn memory(&self) -> &Memory {
+        &self.memory
```
Let's create an issue for this.

plafer replied:
https://github.com/0xPolygonMiden/miden-vm/issues/1613

plafer commented:
@bobbinth @Al-Kindi-0 ready for another look. I addressed all comments, and `TODO(plafer)`. I will open an issue tomorrow for what's left after this PR (namely docs and implementing the additional required range-checks).

bobbinth commented:
> @bobbinth @Al-Kindi-0 ready for another look. I addressed all comments, and `TODO(plafer)`. I will open an issue tomorrow for what's left after this PR (namely docs and implementing the additional required range-checks).

How does the performance look like in this branch vs. `next`?

bobbinth commented on processor/src/errors.rs:
```diff
@@ -129,6 +133,8 @@ pub enum ExecutionError {
       hex = to_hex(.0.as_bytes())
     )]
     SyscallTargetNotInKernel(Digest),
+    #[error("word access at memory address {addr} in context {ctx} is unaligned")]
+    UnalignedMemoryWordAccess { addr: u32, ctx: ContextId },
```
We could still collapse it into a single variant by making `clk` optional - though, not sure if it is worth it.

If we do end up with 2 variants, I would rename this one into something else so that in ends up next to `MemoryUnalignedWordAccess` and would lave a brief comment explaining why we need both of them.

plafer commented on assembly/src/assembler/instruction/mem_ops.rs:
```diff
@@ -125,9 +125,13 @@ pub fn local_to_absolute_addr(
+    validate_param(index_of_local, 0..=max)?;
 
-    push_felt(block_builder, -Felt::from(max - index));
+    // Local values are placed under the frame pointer, so we need to calculate the offset of the
+    // local value from the frame pointer.
+    // The offset is in the range [1, num_proc_locals], which is then subtracted from `fmp`.
+    let fmp_offset_of_local = num_proc_locals - index_of_local;
+    push_felt(block_builder, -Felt::from(fmp_offset_of_local));
```
Yes exactly, it's for word alignment.

> So, in a way, fmp now points to the next available memory cell, right?

Exactly

plafer commented:
> How does the performance look like in this branch vs. next?

No change in the `program_execution` benchmarks

bobbinth commented on processor/src/chiplets/memory/mod.rs:
```diff
@@ -101,25 +109,31 @@ impl Memory {
     /// Unlike read() which modifies the memory access trace, this method returns the value at the
     /// specified address (if one exists) without altering the memory access trace.
-    pub fn get_value(&self, ctx: ContextId, addr: u32) -> Option<Word> {
+    pub fn get_value(&self, ctx: ContextId, addr: u32) -> Option<Felt> {
```
Doc comment (i.e., reference to "word" on line 107) needs to be updated.

plafer replied:
fixed

bobbinth commented on stdlib/asm/crypto/fri/helper.masm:
```diff
@@ -170,17 +171,21 @@ export.load_and_verify_remainder
     # adv_pipe the remainder codeword
     ## Get the length of remainder
-    exec.constants::tmp6 mem_loadw
+    padw exec.constants::tmp6 mem_loadw
```
Was this a bug previously?

Al-Kindi-0 replied:
There was a missing `padw` for sure. Probably became missing when we changed how `reseed` works. The op stack just before that line is empty so nothing is overwritten but still I would've expect some underflow stack error.

bobbinth commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
```diff
@@ -112,39 +112,39 @@ end
 #!
 #! Input: [tau1, tau0, tau_ptr, ...]
-#! Output: [tau_ptr + 513, ...]
+#! Output: [tau_ptr + 513*4, ...]
 #!
 #! Cycles: 8323
 export.powers_of_tau
```
Similar comment as above re cycle count: I think this now takes at least 513 cycles more.

plafer replied:
Updated to ~8850

bobbinth commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
```diff
@@ -154,7 +154,7 @@ end
 export.set_to_zero
     padw
     repeat.512
-        dup.4 add.1 swap.5
+        dup.4 add.4 swap.5
         mem_storew
     end
```
Similar comment as above re updating the cycle count.

plafer replied:
Updated to ~3100

bobbinth reviewed (commented):
Looks good! Thank you! This is almost a full review (I have just a handful of files to go through).

bobbinth commented:
btw, I think we should probably merge this into a new branch - e.g., something like `element-mem`. This way, we can start migrating `miden-base` to this branch but at the same time, would be able to make additional PRs against it.

Al-Kindi-0 commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
Should be doable but will require some tinkering. I can tackle it in a separate PR is that sounds good.

bobbinth replied:
I wonder if we could measure this using an existing test (i.e., by measuring the number of cycles the tests takes to execute) - i think that should put us within 100 cycles of the actuals.

Based on cursory look, it seems like these changes added about 2K cycles to signature verification (not great, not terrible), but would be good to confirm this.

Al-Kindi-0 replied:
Sounds like a good idea, will try that.

Al-Kindi-0 replied:
It seems that the count on the `next` branch was quiet pessimistic. Using the above methodology, here are the cycle counts:

1. `next`: 89032
2. `this`: 90358

plafer replied:
Can we start moving away from commenting cycle counts? Not only is it tedious and a notable time sink, I'm not convinced it's that effective and it's easy to get wrong. Not to mention keeping track of it through branches, loops, etc.

We could instead keep track of cycle count by running `miden analyze` on relevant benchmarks.

bobbinth replied:
Agreed that precise cycle count is not needed in most cases (and probably gets out of sync with reality over time as well). It is useful though to know which procedure are the bottlenecks (e.g., this procedure takes 50K cycles out of 90K for the whole thing).

Not sure what's the best way to reconcile these though. Probably fine to drop cycle counts in most cases, but may be good to keep in some places.

plafer replied:
> It is useful though to know which procedure are the bottlenecks

Totally agree, but I think we should approach this the same way we approach optimizing e.g. Rust programs. That is, use tools to identify bottlenecks. Now in reality, the tools to optimize MASM will not be as powerful than the suite of existing tools for "normal" programs, but nevertheless I'd leave it to whoever is optimizing programs to figure out the bottlenecks on their own (using `miden analyze`, knowledge of the algorithm being optimized, etc).

bobbinth replied:
Yes, this makes sense. Though, since this info is already there, I think we can just make it "less precise". Specifically:

- Let's update the cycle count for the main procedure to `~90400`.
- For the other procedures I commented about let's put approximate cycle counts as well. For example, for `powers_of_tau` we could just estimate it as `~8900`.

plafer replied:
Done

Al-Kindi-0 commented on air/src/constraints/chiplets/memory/mod.rs:
```diff
@@ -123,47 +126,108 @@ fn enforce_d_inv<E: FieldElement>(
+    result[6] = memory_flag_no_last * binary_not(frame.f_scw_next()) * c2 * frame.v_next(2);
+    result[7] = memory_flag_no_last * binary_not(frame.f_scw_next()) * c3 * frame.v_next(3);
+
+    // non-first row, same word or context constraints: when row' is in the same word/ctx, and
```
should it be `same word and context`?

plafer replied:
Yes, fixed. I also switched to "word address"

Al-Kindi-0 commented on air/src/constraints/chiplets/memory/mod.rs:
```diff
@@ -174,16 +238,36 @@ fn enforce_values<E: FieldElement>(
+    fn is_word_access_next(&self) -> E;
     /// The current context value.
     #[allow(dead_code)]
     fn ctx(&self) -> E;
     /// The current address.
     #[allow(dead_code)]
-    fn addr(&self) -> E;
+    fn word_next(&self) -> E;
```
should we remove these if they are not used?

plafer replied:
Agreed, removed

Al-Kindi-0 commented on air/src/constraints/chiplets/memory/mod.rs:
```diff
@@ -174,16 +238,36 @@ fn enforce_values<E: FieldElement>(
+    /// The 1st bit of the index of the memory address in the next word.
+    fn idx1_next(&self) -> E;
     /// The current clock cycle.
     #[allow(dead_code)]
     fn clk(&self) -> E;
```
same comment here

plafer replied:
Removed

Al-Kindi-0 commented on air/src/constraints/chiplets/memory/mod.rs:
```diff
@@ -221,38 +309,34 @@ trait EvaluationFrameExt<E: FieldElement> {
     /// The difference between the next context and the current context.
     fn ctx_change(&self) -> E;
     /// The difference between the next address and the current address.
```
nit: word address

plafer replied:
fixed

Al-Kindi-0 reviewed (commented):
Went through the constraints and all looks good except for a few questions and comments.

plafer commented on air/src/constraints/chiplets/memory/mod.rs:
```diff
@@ -123,47 +126,108 @@ fn enforce_d_inv<E: FieldElement>(
+///   being written need to be equal to the values in the previous row (i.e. previously written, or
+///   initialized to 0).
+///   - The implication is that in a given evaluation frame, we always constrain the "next" value,
+///     since that constraint depends on the "current" value.
```
To know the constraint for a value, we have to look at the previous row. So to be able to look at the "previous" row in a constraint, the constraint can only apply to the "next" row in the frame (where its previous row is the "current" one).

It's a bit hard to put in words, and I don't know that it added value, so I just removed it.

plafer commented on air/src/constraints/chiplets/memory/mod.rs:
```diff
@@ -123,47 +126,108 @@ fn enforce_d_inv<E: FieldElement>(
+        let f3 = frame.idx1_next() * frame.idx0_next();
+
+        let c_i = |f_i| {
+            // z_i is set to 1 when `v'[i]` is not being accessed.
```
>  but not the i-th element

It's not quite the i-th element that's not set to 1; doesn't that mean that all `i` are not set to 1?

I reworded to something that's hopefully better

Al-Kindi-0 replied:
> It's not quite the i-th element that's not set to 1; doesn't that mean that all i are not set to 1?

Yeah it is the z_i we are talking about.
I still feel that the sentence doesn't capture the function of z_i but it should be easy to get it from the expression in the second line, hopefully.

plafer replied:
I updated it to be your suggestion `z_i is set to 1 when we are operating on elements but not the i-th element`

plafer commented on assembly/src/assembler/mod.rs:
```diff
@@ -567,19 +567,23 @@ impl Assembler {
         // Make sure the current procedure context is available during codegen
         let gid = proc_ctx.id();
+
+        // We expect the number of locals to be a multiple of the word size, having been rounded up
+        // if necessary.
         let num_locals = proc_ctx.num_locals();
+        assert_eq!(num_locals % WORD_SIZE as u16, 0);
```
The converse is the case where the user declares `proc.my_fn.1`, and writes a word with `loc_storew.0`. Given what's in the `.masm` file, this is incorrect since only 1 element was allocated (but 4 are being written). But this assembles and runs fine, because of our padding. So I opted for using the padded number of locals everywhere as it is the true number of locals (and even in the error reporting if any), and documenting the padding behavior (in #1616).

So the way to think about `proc.my_fn.1` is that the user *suggests* a number of locals to be allocated, but the assembler is free to allocate more. I see this as the downside of allowing the user to not have to write the "padded" number of locals.

Or maybe that's more confusing than what you're suggesting?

bobbinth replied:
> The converse is the case where the user declares `proc.my_fn.1`, and writes a word with `loc_storew.0`. Given what's in the `.masm` file, this is incorrect since only 1 element was allocated (but 4 are being written).

Shouldn't this be an assembly-time error? If I'm understanding correctly, in the current approach, this won't trigger the error (because the assembler would think that the number of locals is 4). In the approach I've suggested, it would be an error because the assembler would know that the number of locals is 1.

plafer replied:
Ah, I thought our goal with rounding up was to *not* output an error in this case. So then we would round up, but still ensure that what was written by the user "makes sense" (i.e. it's ok to do `loc_store.0`, but not `loc_storew.0` in `proc.my_fn.1`, even though internally we're rounding up)?

bobbinth replied:
I think the main reason for rounding up is to ensure that locals start at a word-aligned address. If not for this, we probably wouldn't need to round up.

So, the way I see it, from the user's standpoint everything should work as if there is no rounding up. So, if they declare a procedure with only 1 local, but then try to do `loc_storew.0`, that should be an error.

plafer replied:
Makes sense. Fixed

plafer commented on assembly/src/assembler/procedure.rs:
```diff
@@ -55,6 +58,15 @@ impl ProcedureContext {
+    // value + 3 & !3 = 4,8,8,8 (&!3 clears the last two bits)
+    // as desired.
+    (value + 3) & !3
+}
```
Ah didn't know about that one. Fixed

plafer commented on assembly/src/assembler/procedure.rs:
```diff
@@ -44,8 +44,11 @@ impl ProcedureContext {
+    /// The number of locals is rounded up to the nearest multiple of 4.
     pub fn with_num_locals(mut self, num_locals: u16) -> Self {
-        self.num_locals = num_locals;
+        self.num_locals = round_up_to_multiple_of_4(num_locals);
```
Changed to use `next_multiple_of` for now

plafer commented on assembly/src/errors.rs:
```diff
@@ -61,6 +61,15 @@ pub enum AssemblyError {
     },
+    #[error("invalid local word index: {local_addr}")]
+    #[diagnostic(help("the index to a local word must be a multiple of 4"))]
+    InvalidLocalWordIndex {
+        #[label]
+        span: SourceSpan,
+        #[source_code]
+        source_file: Option<Arc<SourceFile>>,
+        local_addr: u16,
+    },
```
Yes, this is the error that occurs when you write e.g. `loc_loadw.3` or `loc_storew.3` (since 3 is not word-aligned).

plafer commented on assembly/src/tests.rs:
```diff
@@ -842,7 +842,7 @@ fn mem_operations_with_constants() -> TestResult {
     const.GLOBAL_LOADW_PTR={GLOBAL_LOADW_PTR}
 
-    proc.test_const_loc.4
+    proc.test_const_loc.24
```
Ah, actually 12 is needed, since we run a `loadw` at local address 8 (so addresses 8-11 are accessed). Fixed

plafer commented on processor/src/chiplets/aux_trace/mod.rs:
```diff
@@ -970,11 +968,15 @@ where
-fn build_value<E: FieldElement<BaseField = Felt>>(alphas: &[E], elements: &[Felt]) -> E {
-    assert_eq!(alphas.len(), elements.len());
+#[inline(always)]
+fn build_value<E: FieldElement<BaseField = Felt>, const N: usize>(
```
I simplified the entire docstring

plafer commented on stdlib/tests/mem/mod.rs:
```diff
@@ -38,86 +38,87 @@ fn test_memcopy() {
             exec.sys::truncate_stack
         end",
-        mem_addr
+        end_addr, start_addr,
```
fixed

plafer commented on stdlib/asm/crypto/stark/constants.masm:
```diff
@@ -5,9 +5,17 @@
+
+# Number of coefficients corresponds to "number of main & aux columns" + 8,
+# where "8" is the number of columns needed to store the constraint composition polynomial.
 const.NUM_DEEP_COMPOSITION_COEF_MULTIPLIED_BY_TWO_ROUNDED_UP_TO_FOUR=88
```
I think so yes, fixed

bobbinth reviewed (approved):
Looks good! Thank you! I didn't review the constraints in too much detail, but will do this when reviewing the docs PR.

No additional comments from this PR, but there are a couple of comments remain from the previous review (re handling padding for locals and updating cycles counts so that at least what we have there is not wrong).

Al-Kindi-0 commented:
All looks good to me now, thanks again!

plafer commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
```diff
@@ -76,31 +76,31 @@ end
 #! Output: [...]
 #!
 #! Cycles: 1327
-export.hash_to_point.2
```
It's [2 cycles](https://github.com/0xPolygonMiden/miden-vm/blob/868bf1e78567539d2b1ae4cc42bb5b8edd788c9e/assembly/src/assembler/instruction/field_ops.rs#L52), not 4, right? 

Anyhow I updated the cycle count to be ~1400

bobbinth replied:
Ah yes - that's a typo. 2 cycles is correct.

plafer commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
```diff
@@ -545,7 +545,7 @@ export.compute_s1_norm_sq
         swap
 
         # 6) Increment the pointer
-        add.1
```
Updated to ~59000

plafer commented on stdlib/asm/crypto/dsa/rpo_falcon512.masm:
```diff
@@ -565,7 +565,7 @@ end
 export.compute_s2_norm_sq
     repeat.128
         padw
-        dup.4 add.1 swap.5
```
Updated to ~13500
